### PR TITLE
Azure Retail API Support + Other Features

### PR DESCRIPTION
**Pull request by DanielMabbett**
## Incoming Changes

* Added .devcontainer feature to allow developers to work with in a local environment quicker. Spins up a db and app container and shells into the app to get started with dev work. 
* Added Azure Retail Pricing API
* Updated Node JS Version to 14.15.0 `may be breaking change!`
* Updated MongoDB Version to 4.4 `may be breaking change!`
* Updated package.json 
* Added example docker-compose to bring up dockerfile with a mongo db container for running
* Added some example queries in a new folder
* Updated Readme

### CODE REVIEW

File: src/db/helpers.ts [hunk 1]
```diff
@@ -7,7 +7,11 @@ function generateProductHash(product: Product): string {
   // keep AWS product hashes the same so Infracost tests don't break
   if (product.vendorName === 'aws') {
     hashFields = ['vendorName', 'sku'];
-  } else {
+  } 
+  else if (product.vendorName === 'azure') {
```
Comment (aliscott) [inline src/db/helpers.ts:11]: I don't think we need a specific condition for this. The one's for AWS are only there so we didn't affect old data we had.

File: src/db/helpers.ts [hunk 2]
```diff
@@ -29,7 +33,17 @@ function generatePriceHash(product: Product, price: Price): string {
       'termPurchaseOption',
       'termOfferingClass',
     ];
-  } else {
+  }
+  else if (product.vendorName === 'azure') {
```
Comment (aliscott) [inline src/db/helpers.ts:37]: Same as above

File: src/updaters/azureRetail.ts [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+function parsePrices(product: Product, productJson: ProductJson): Price[] {
+    const prices: Price[] = [];
+
+            const price: Price = {
+                priceHash: '',
+                purchaseOption: productJson.type,
+                unit: productJson.unitOfMeasure,
+                USD: `${productJson.unitPrice}`,
+                effectiveDateStart: productJson.effectiveStartDate,
+                startUsageAmount: productJson.retailPrice
```
Comment (aliscott) [inline src/updaters/azureRetail.ts:143]: This should be the usage amount required before the price kicks in, so for Azure I think this means the `tierMinimumUnits`.

File: src/updaters/azureRetail.ts [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    const product: Product = {
+        productHash: '',
+        sku: productJson.skuId,
+        vendorName: 'azure',
+        region: productJson.location,
+        service: productJson.serviceName,
+        productFamily: productJson.serviceFamily,
+        attributes: {
+            type: productJson.type
+        },
```
Comment (aliscott) [inline src/updaters/azureRetail.ts:124]: We should have some more attributes here. My philosophy here has basically been to add everything we can get.

For example for looking up the price of a Terraform azure `azurerm_virtual_machine` it looks like the `vm_size` attribute maps directly to the value of `armSkuName` provided by the Azure retail API. So we would want that in here so we can query by it.

**Review by aliscott — COMMENTED** [review]
Thanks @DanielMabbett this is a great addition! Thanks for improving the dev experience as well.

I haven't managed to do a deep dive into Azure pricing but have left an initial review.

A couple of general notes:

1. When I run the Azure updater I only see 77 Azure products added to the Database. I think this is partly due to not following the `NextPageLink`, but even with that considered I'd expect to see 100.

2. It would be good to validate that we can use the pricing API for Infracost resources. One way of doing this would be to add a sample [azurerm_virtual_machine](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/virtual_machine) resource to Infracost (docs: https://github.com/infracost/infracost/blob/master/CONTRIBUTING.md#adding-new-resources) and test it against this branch.

File: src/db/helpers.ts (hunk 1 above)
Comment (DanielMabbett) [reply src/db/helpers.ts]: Removed this :)

File: src/db/helpers.ts (hunk 2 above)
Comment (DanielMabbett) [reply src/db/helpers.ts]: Removed this :)

**Pull request merged**